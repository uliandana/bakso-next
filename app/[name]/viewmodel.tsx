import BerryAPIDataSourceImpl from '@/Data/DataSource/API/BerryAPIDataSource';
import PokemonAPIDataSourceImpl from '@/Data/DataSource/API/PokemonAPIDataSource';
import { BerryRepositoryImpl } from '@/Data/Repository/BerryRepositoryImpl';
import { PokemonRepositoryImpl } from '@/Data/Repository/PokemonRepositoryImpl';
import { Berry } from '@/Domain/Model/Berry';
import { Pokemon } from '@/Domain/Model/Pokemon';
import { GetBerry } from '@/Domain/UseCase/Berry/GetBerry';
import { GetPokemonByName } from '@/Domain/UseCase/Pokemon/GetPokemonByName';
import { useEffect, useState } from 'react';
import { getStoredBerry, storeBerry } from './indexeddb';

export default function NameViewModel(name: string) {
  const [pokemon, setPokemon] = useState<Pokemon>();
  const [berries, setBerries] = useState<Berry[]>([]);
  const [iseFetching, setIsFetching] = useState<Boolean>(true);

  const pokemonDataSourceImpl = new PokemonAPIDataSourceImpl();
  const pokemonRepositoryImpl = new PokemonRepositoryImpl(pokemonDataSourceImpl);

  const getPokemonUseCase = new GetPokemonByName(pokemonRepositoryImpl);

  const berryDataSourceImpl = new BerryAPIDataSourceImpl();
  const berryRepositoryImpl = new BerryRepositoryImpl(berryDataSourceImpl);

  const getBerryUseCase = new GetBerry(berryRepositoryImpl);

  const fetchPokemon = async () => {
    try {
      setIsFetching(true);
      const data = await getPokemonUseCase.invoke(name);
      setIsFetching(false);
      if (data) {
        setPokemon(data);
      }
    } catch(e) {
      setIsFetching(false);
    }
  };

  const fetchBerry = async () => {
    try {
      const data = await getBerryUseCase.invoke();
      if (data) {
        setBerries(data);
        storeBerry(data);
      }
    } catch(e) {
      setBerries([]);
    }
  };

  useEffect(() => {
    fetchPokemon();
    fetchBerry();
  }, []);

  useEffect(() => {
    if (berries.length) {
      const cb: (b:Berry[]) => Berry[] = b => b;
      getStoredBerry(cb);
    }
  }, [berries]);

  return {
    pokemon,
    berries,
    iseFetching,
  };
}
