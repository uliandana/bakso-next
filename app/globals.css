@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  --foreground-rgb: 255, 255, 255;
  --background-start-rgb: 188,55,54;
  --background-end-rgb: 188,55,54;
  font-size: calc(768px * 16 / 375);
}

@media (prefers-color-scheme: dark) {
  :root {
    --foreground-rgb: 255, 255, 255;
    --background-start-rgb: 0, 0, 0;
    --background-end-rgb: 0, 0, 0;
  }
}

@layer utilities {
  .card-flip {
    position: absolute;
    backface-visibility: hidden;
    transform: rotateY(0);
    transition: transform 0.5s ease-in-out;
  }
  .card-flipped {
    position: absolute;
    backface-visibility: hidden;
    transform: rotateY(180deg);
    transition: transform 0.5s ease-in-out;
  }
  .silhouette {
    filter: brightness(0) drop-shadow(0.0625rem 0.125rem 0.25rem rgb(var(--background-start-rgb)));
  }
}

body {
  color: rgb(var(--foreground-rgb));
  background: linear-gradient(
      to bottom,
      transparent,
      rgb(var(--background-end-rgb))
    )
    rgb(var(--background-start-rgb));
}

* {
  animation: fade-in 0.5s ease-in-out;
}

@keyframes loading {
  to { transform: translateX(100%); }
}

@keyframes fade-in {
  from { opacity: 0; }
  to { opacity: 1; }
}

@media only screen and (max-width: 1440px) {
  :root {
    font-size: calc(100vw * 16 / 1440);
  }
}

@media only screen and (max-width: 768px) {
  :root {
    font-size: calc(100vw * 16 / 768);
  }
}
