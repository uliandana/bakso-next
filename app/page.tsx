'use client'
import useViewModel from './viewmodel';
import Header from './_components/header.root';
import List from './_components/list.root';
import ButtonChoose from './_components/buttonchoose.root';
import styles from './styles.module.css';
import ScrollTop from './_components/scrolltop.root';

export default function Root() {
  const { pokemons, select, onChoosePokemon, isFetching, search, scrollTop } = useViewModel();

  return (
    <main className={styles.main} ref={scrollTop.ref}>
      <Header search={search} />
      <List pokemons={pokemons} isFetching={isFetching} select={select} />
      <ButtonChoose select={select} onChoosePokemon={onChoosePokemon} />
      <ScrollTop scrollTop={scrollTop} />
    </main>
  )
}
