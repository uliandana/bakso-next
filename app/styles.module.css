.main {
  @apply m-auto h-dvh overflow-y-auto;
}

.main > header {
  @apply p-[2rem] px-[6rem];
}

.main > header > h1 {
  @apply text-[3rem] font-bold text-center mb-[2rem];
}

.main > header > input {
  @apply block w-full text-[2rem] text-black p-[1rem] bg-white rounded-[1rem] border-solid border-[0.0625rem] border-slate-500;
}

.main > ul {
  @apply grid grid-cols-4;
}

.item {
  @apply h-[25vw] text-center flex flex-col items-center justify-between overflow-hidden relative cursor-pointer;
}

.item > img {
  @apply block w-full transition-all hover:drop-shadow-xl hover:rotate-6 hover:scale-110;
}

.item > p {
  @apply text-[1.5rem] md:text-[2rem] uppercase font-[700] absolute bottom-1;
}

.loading {
  @apply block p-[1rem] mb-[1rem] text-center;
}

.main > footer {
  @apply sticky bottom-[-3rem] mb-[-3rem] py-[6rem] px-[10rem];
}

.main > footer > button {
  @apply w-full py-[1rem] rounded-[3rem] text-[2rem] uppercase bg-neutral-100 text-red-600 font-[700] tracking-[0.125rem] shadow-2xl;
}

.btnScrollTop {
  @apply sticky bottom-[3.5rem] right-[3rem] flex ml-auto p-[0.5rem] size-[4rem] rounded-full uppercase bg-neutral-100 text-red-600 shadow-2xl;
}
