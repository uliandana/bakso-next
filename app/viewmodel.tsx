import { useEffect, useState } from 'react';
import PokemonAPIDataSourceImpl from '@/Data/DataSource/API/PokemonAPIDataSource';
import { PokemonRepositoryImpl } from '@/Data/Repository/PokemonRepositoryImpl';
import { GetPokemon } from '@/Domain/UseCase/Pokemon/GetPokemon';
import { Pokemon } from '@/Domain/Model/Pokemon';

export default function RootViewModel() {
  const [pokemons, setPokemons] = useState<Pokemon[]>([]);
  const [observer, setObserver] = useState<IntersectionObserver | null>(null);
  const [indexPage, setIndexPage] = useState<number>(0);
  const [offset, setOffset] = useState<number>(0);
  const [isFetching, setIsFetching] = useState<Boolean>(true);

  const pokemonDataSourceImpl = new PokemonAPIDataSourceImpl();
  const pokemonRepositoryImpl = new PokemonRepositoryImpl(pokemonDataSourceImpl);
  
  const getPokemonUseCase = new GetPokemon(pokemonRepositoryImpl);

  const fetchPokemon = async (offset: number) => {
    try {
      setIsFetching(true)
      const data = await getPokemonUseCase.invoke(offset);
      setPokemons([...pokemons, ...data]);
      setIsFetching(false);
    } catch(e) {
      setPokemons([...pokemons]);
      setIsFetching(false);
    }
  };

  const initializeObserver = () => {
    const options = {
      threshold: 1,
    };
    
    const fnObs = new IntersectionObserver((e) => {
      const { isIntersecting, target } = e[0];
      if (!isIntersecting) {
        return;
      }

      const newOffset = parseInt(target.getAttribute('data-pokemon') || '0');
      if (!newOffset) {
        return;
      }
      
      setOffset(newOffset);
    }, options);

    setObserver(fnObs);
  };

  useEffect(() => {
    const nextIndexPage = Math.ceil(offset / 100);
    if (indexPage >= nextIndexPage) {
      return;
    }

    setIndexPage(nextIndexPage);
    fetchPokemon(offset);
  }, [offset]);

  useEffect(() => {
    const target = document.querySelector(`[data-pokemon='${(indexPage + 1) * 100}']`);
    if (target && observer) {
      observer.observe(target);
    }
  }, [pokemons]);

  useEffect(() => {
    fetchPokemon(0);
    initializeObserver();
  }, []);

  return {
    pokemons,
    isFetching
  };
}
