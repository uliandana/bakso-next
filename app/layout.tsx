import type { Metadata } from 'next'
import { Inter } from 'next/font/google'
import Script from 'next/script'
import './globals.css'

const inter = Inter({ subsets: ['latin'] })

export const metadata: Metadata = {
  title: 'Bakso',
  description: 'Feed your hunger',
}

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode
}>) {
  return (
    <html lang="en">
      <body className={inter.className}>{children}</body>
      <Script src="/sw-install.js" />
    </html>
  )
}
