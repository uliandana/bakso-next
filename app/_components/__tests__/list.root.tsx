import React from 'react';
import { describe, expect, jest, test } from '@jest/globals';
import { Pokemon } from '@/app/viewmodel';
import { createRenderer } from 'react-test-renderer/shallow';
import List from '../list.root';

const dummyPokemon: Pokemon = {
  id: '10',
  nameSlug: 'pikachu',
  name: 'Pikachu',
  sprite: 'url',
  weight: 100,
  baseWeight: 100,
  bgColor: '#fff',
  stats: [],
  types: ['bug'],
  evolvesTo: [],
};

describe('./app/_components/list.root', () => {
  test('renders', () => {
    const select = {
      value: 'pikachu',
      onSelectCard: jest.fn(),
    };
    const tree = createRenderer().render(<List pokemons={[dummyPokemon]} isFetching={false} select={select} />);
    expect(tree).toMatchSnapshot();
  });

  test('no border when pokemon not selected', () => {
    const select = {
      value: 'mew',
      onSelectCard: jest.fn(),
    };
    const res = List({
      pokemons: [dummyPokemon],
      isFetching: false,
      select,
    });
    const label = res.props.children[0][0];
    expect(label.props.style.border).toBe('');
  });

  test('trigger event when pokemon selected', () => {
    const select = {
      value: 'mew',
      onSelectCard: jest.fn(),
    };
    const res = List({
      pokemons: [dummyPokemon],
      isFetching: false,
      select,
    });
    const label = res.props.children[0][0];
    const input = label.props.children[0];
    input.props.onChange();
    expect(select.onSelectCard).toBeCalledWith('pikachu');
  });
});
