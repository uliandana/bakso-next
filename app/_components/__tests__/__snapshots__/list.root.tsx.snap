// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`./app/_components/list.root renders 1`] = `
<div
  className="list"
>
  <label
    data-pokemon={1}
    htmlFor="pokemon-10"
    style={
      {
        "backgroundColor": "#fff",
        "border": "0.0625rem solid gold",
      }
    }
  >
    <input
      className="hidden"
      id="pokemon-10"
      name="berry"
      onChange={[Function]}
      type="radio"
      value="pikachu"
    />
    <img
      alt="Pikachu"
      className="block w-full transition-all hover:drop-shadow-xl hover:rotate-6 hover:scale-110"
      src="url"
    />
    <p>
      Pikachu
    </p>
  </label>
</div>
`;
